### ISSUE TITLE
Game logic systems

### ISSUE THREAD

**Issue by marrub--** [Team-Doominati/Doominati#34]
We need a lot of thought for how the whole of logic should work.
This is more of an open question that should lead both to discussion here and via new issues.

Here's some main components of game logic, from lowest to highest level:
- Framerate (game snapshot rate)
- Positioning systems (how big is x/y/z? how do we know? can we change it? how does this change rendering?)
- Thinkers (objects which call functions during game snapshot and store data)
- Large, sectioned collision data and graphics (maps)
- User input
- Rendering occlusion

And some parts of thinkers we need to address (also in order of lowest to highest level):
- Data mapping and storage
- Callbacks (what functions do we call and when?)
- Events (how do we trigger what, when and where and why?)
- Ray intersections (hitscans)
- Collision response
- Physics
- Animation (discussed elsewhere, but recap and ideas would be good)
- AI

Also important, I feel, is how we name things.
When we add something, we're going to be typing it and reading it a lot, so let's make sure we're happy with how we name it.

**Comment by Kyle873** [comment]
- Framerate: The only reason I see the need to adjust this would be for effects such as time dilution and the like, which would certainly be a nice thing to have.
- Positioning: Should we provide a native to let the user setup the coordinate system and the distance/ratio of coordinate movements or keep those measurements static? How much of a pain will that be to keep together on the actual engine side of things would be my concern with that.
- Maps: We have touched on this a few times, but ultimately I remember deciding that the map system itself would be done by scripting and the user as opposed to a static format provided by the engine for maximum flexibility. Whether we still want to take this approach or not, I'm not sure.
- Input: We've discussed it in Discord, and I think we've come up with how we want to do it now. I know @marrub-- has a pretty solid idea of how he wants to approach it now anyway, at least.

As for the thinker stuff, we should probably discuss that in Discord and leave key points here, otherwise this comment section will turn into an absolute shitstorm.

**Comment by marrub--** [comment]
- Framerate

I think 50 gametics per second is good and that with rendering interpolation (already implemented as argument to Draw callbacks) looks and plays quite smooth.

Being able to adjust it would really only ever be used for, as you said, changing the time-scale.

- Positioning

You gave me an interesting idea.

Much like how most vector math libraries have a way to set the "up" vector, it would be a good idea to have a function for setting up dimensions, like:

```C
/* syntax: xfacing, yfacing, zfacing */
DGE_SetupNormal(DGE_Right, DGE_Up, DGE_Forward);
```

So you can use any arbitrary system.

On the engine side, this should not be a problem because we scarcely use actual coordinate calculations.

Most of that is done by user code.

- Maps

Still not sure, we'll touch on that when we reach it I suppose.

- Input

The idea I have is, I think, pretty solid. It's also already implemented (mostly.)

It's to be set up so there will be a fixed number of actual buttons (which do actions) and actual axes (which do movement etc.)

This should be *very* good for sync-based networking, as Doom's model has proven over the years.

Currently, it is limited by my hasty implementation, but it does work for now.

Input sources can be from network pipes, demos, or just the program's event pump.

- Thinkers

heheh

We've decided on this kind of hierarchy and naming scheme:

> Thinker
>> PositionThinker
>> RenderThinker
>> PhysicsThinker
>>> Entity

Where Entity is the most-derived class from the Thinker base, and can be used for any real, tangible game entity; and Thinker (and its derivitives) should be used for more abstract data-based game operations.

**Comment by DavidPH** [comment]
Framerate: I would also like for it to be adjustable. Should be reasonably simple to do so, but we will need to consider the implications for input handling with respect to net code.

Positioning: My original idea for coordinates was s24.7 fixed point where 1 would correspond to pixels. Mind you, this originates from MgC, where render sizes were fixed and a larger window showed more of the game world. The idea being that this would ensure that the limited number of fractional bits would inherently be enough, since even zoomed in they would represent sub-rendered distances. For other game types where the visible game-space is fixed (so a larger windows results in things being drawn larger), this might not really be sufficient.

Thinkers: As Marrub said, this has mostly been decided. I will add a note that the base Thinker class is visible from scripting, and I expect most of its derived types to be creatable from scripts.

Maps: I still expect to do external map formats using pure scripting. This could be done by way of a base scripting library project once we have something reasonably settled.

Input: We have a basic model set up that should be sufficient for everything we need. Just need to fill it out with enough distinct virtual buttons and such. Also need keybinding and netcode. But all of those should be possible to implement on top of what we have.

**Comment by marrub--** [comment]
Besides the facing of each position (which is as simple as multiplying something by a normal vector given by a script) really all it comes down to is how much precision we have and how many dimensions.

There should easily be 3 dimensions (surprisingly useful in 2d games), and 24.7... should suffice.

While I would like floating point or a 64bit type, these have drawbacks that are too large.

For maps, I'm less concerned about the format (doing that with scripts is :+1:) than I am about the ways that they're handled.

Logic occlusion, area sets, visible sets, spatial subdivision (BSP, portal, kd-tree, quadtree, etc) which **need** to be in-engine for optimization, handling collides for lighting, and maaaaaaaaaaany more questions arise.

**Comment by DavidPH** [comment]
I agree on wanting three dimensions. And remember that s24.7 can be a storage/transfer type even if we use a different evaluation type for intermediaries. Using a 32-bit fixed-point type has the benefit of clearly defining the bounds of the world, allowing for a logical basis from which to choose evaluation types.

**Comment by marrub--** [comment]
After lots of thonkang and code emittance and blood sacrifices, we have come to conclusions on most of these:
- Framerate (game snapshot rate)
Locked at 50. Possible native or configuration of sorts later for this.

- Positioning systems (how big is x/y/z? how do we know? can we change it? how does this change rendering?)
Positioning is 24.8 fixed point, right-down forward vector and cannot be changed as the internal physics assume this. Rendering this is up to the user using matrices and size members.

- Thinkers (objects which call functions during game snapshot and store data)
**Delicious**.

- Large, sectioned collision data and graphics (maps)
***Delicious***. Do it with scripts.

- User input
Not done, but API has been discussed elsewhere.

- Rendering occlusion
yappa shiran lol

And some parts of thinkers we need to address (also in order of lowest to highest level):
- Data mapping and storage
***Fuckin' umai m8***

- Callbacks (what functions do we call and when?)
I think David has plans for this but I'm not sure

- Events (how do we trigger what, when and where and why?)
Same as callbacks. We have some global ones already. The rest is inputs. No interrupts.

- Ray intersections (hitscans)
I think David made a ray intersection function.

- Collision response
Done.

- Physics
Done.

- Animation (discussed elsewhere, but recap and ideas would be good)
Done via scripts.

- AI
Done via scripts.

And finally, as for naming, we've decided on and propogated a very consistent naming scheme for natives.

STATUS OVERVIEW
POINTS EARNED: 99995153251414/100200400204002040
AREAS COMPLETED: 10/14
FINGERLESS GLOVE%: 99
